lib/routes.js

```diff
@@ -7,49 +7,93 @@ const { formatError, GraphQLError } = require('graphql')
 const subscription = require('./subscription')
 const { FEDERATED_ERROR, toGraphQLError } = require('./errors')
 
-const responseSchema = {
-  '2xx': {
-    type: 'object',
-    properties: {
-      data: {
-        type: ['object', 'null'],
-        additionalProperties: true
-      },
-      errors: {
-        type: 'array',
-        items: {
-          type: 'object',
-          required: ['message'],
-          properties: {
-            message: { type: 'string' },
-            locations: {
-              type: 'array',
-              items: {
-                type: 'object',
-                properties: {
-                  line: { type: 'integer' },
-                  column: { type: 'integer' }
-                }
-              }
-            },
-            path: {
-              type: 'array',
-              items: { type: 'string' }
-            },
-            extensions: {
-              type: 'object',
-              properties: {
-                code: { type: 'string' }
-              },
-              additionalProperties: true
+const responseProperties = {
+  data: {
+    type: ['object', 'null'],
+    additionalProperties: true
+  },
+  errors: {
+    type: 'array',
+    items: {
+      type: 'object',
+      required: ['message'],
+      properties: {
+        message: { type: 'string' },
+        locations: {
+          type: 'array',
+          items: {
+            type: 'object',
+            properties: {
+              line: { type: 'integer' },
+              column: { type: 'integer' }
             }
           }
+        },
+        path: {
+          type: 'array',
+          items: { type: 'string' }
+        },
+        extensions: {
+          type: 'object',
+          properties: {
+            code: { type: 'string' }
+          },
+          additionalProperties: true
         }
       }
     }
   }
 }
 
+const requestProperties = {
+  query: {
+    type: 'string'
+  },
+  persisted: {
+    type: 'boolean'
+  },
+  operationName: {
+    type: 'string'
+  }
+}
+
+const getSchema = {
+  querystring: {
+    type: 'object',
+    properties: {
+      ...requestProperties,
+      variables: {
+        type: 'string' // Stringified JSON
+      }
+    }
+  },
+  response: {
+    '2xx': {
+      type: 'object',
+      properties: responseProperties
+    }
+  }
+}
+
+const postSchema = (allowBatchedQueries) => ({
+  body: {
+    type: allowBatchedQueries ? ['object', 'array'] : 'object',
+    properties: {
+      ...requestProperties,
+      variables: {
+        type: ['object', 'null']
+      }
+    }
+  },
+  // JSON schema isn't allowing ['object', 'array'] on response.
```

Comment:
Have you tried using an `anyOf`? I would prefer to keep the speedup done by the schema.

Reply:
I've had a decent look into this after your suggestion, and I can't seem to make it work. Any combination of `anyOf` I tried I got errors compiling the schema.

If you have further suggestions as to where to look for a possible example or e.g. where in the `fast-json-stringify` codebase I might get a lead on what wall I'm hitting, I'm happy to look into it more - am not particularly familiar with JSON Schema though.

Reply:
Let's land this as is then, we'll investigate in the future.